prov/gni/src/gnix_msg.c

```diff
@@ -558,6 +728,315 @@ static int __gnix_rndzv_req(void *arg)
+						/* We'll try again. */
+						GNIX_INFO(FI_LOG_EP_DATA,
+							  "_gnix_nic_tx_alloc()"
+							  " returned %s\n",
+							  fi_strerror(-ret));
+						return -FI_ENOSPC;
+					}
+
+					txd->completer_fn =
+						__gnix_rndzv_iov_req_complete;
+					txd->req = req;
+					txd->gni_desc.type = GNI_POST_FMA_GET;
+					txd->gni_desc.cq_mode =
+						GNI_CQMODE_GLOBAL_EVENT;
+					txd->gni_desc.dlvr_mode =
+						GNI_DLVMODE_PERFORMANCE;
+					txd->gni_desc.local_mem_hndl =
+						req->msg.recv_info[j].
+						mem_hndl;
+
+					/* TODO: If the user passes in a
+					 * incorrect mdh can we report a
+					 * useful error? */
+					txd->gni_desc.remote_mem_hndl =
+						req->msg.send_info[i].mem_hndl;
+					txd->gni_desc.rdma_mode = 0;
+					txd->gni_desc.src_cq_hndl =
+						(use_tx_cq_blk) ?
+						nic->tx_cq_blk : nic->tx_cq;
+
+					/* TODO: handle alignment! */
+					txd->gni_desc.local_addr =
+						(uint64_t) recv_ptr;
+					txd->gni_desc.remote_addr =
+						req->msg.send_info[i].send_addr;
+					txd->gni_desc.length = get_len;
+					ct_size += get_len;
+
+					next_ct = &txd->gni_desc.next_descr;
+				} else {
+					cur_ct = *next_ct = malloc(sizeof(gni_ct_get_post_descriptor_t));
+
+					cur_ct->ep_hndl = gni_ep;
+					cur_ct->length = get_len;
+					ct_size += get_len;
+					cur_ct->remote_addr = req->msg.send_info[i].send_addr;
+					cur_ct->remote_mem_hndl =
+						req->msg.send_info[i].mem_hndl;
+					cur_ct->local_addr =
+						(uint64_t) recv_ptr;
+					cur_ct->local_mem_hndl =
+						req->msg.recv_info[j].
+						mem_hndl;
+
+					next_ct = &cur_ct->next_descr;
+				}
+			}
+		}
+
+		/* Update the local and remote addresses */
+		recv_len -= get_len;
+
+		/* We have exhausted the current recv (and possibly send)
+		 * buffer */
+		if (recv_len == 0) {
+			j++;
+
+			/* We cannot receive any more. */
+			if (j == recv_cnt)
+				break;
+
+			recv_ptr = req->msg.recv_info[j].recv_addr;
+			recv_len = req->msg.recv_info[j].recv_len;
+
+			/* Also exhausted send buffer */
+			if (get_len == req->msg.send_info[i].send_len) {
+				i++;
+			} else {
+				req->msg.send_info[i].send_addr += get_len;
+			}
+		} else {	/* Just exhausted current send buffer. */
+			i++;
+			recv_ptr += get_len;
+		}
+		GNIX_DEBUG(FI_LOG_EP_DATA, "i = %d, j = %d\n", i, j);
+	}
+
+	/*
+	 * If we ran out of buffer space on the sender's/receiver's side in the
+	 * middle of building the ct, we must terminate and add that ct to the
+	 * queue. Note that if the last txd built was a rdma txd then the txd
+	 * will have been queued and txd will have a NULL value.
+	 */
+	if (txd) {
+		GNIX_DEBUG(FI_LOG_EP_DATA, "\n");
+		*next_ct = NULL;
+		txd_cnt++;
+		slist_insert_tail(&txd->slist, &req->iov_txd_slist);
+	}
+
+	atomic_set(&req->msg.outstanding_txds, txd_cnt);
+
+	/* All the txd's are built, update the work_fn in case of failed post */
+	req->work_fn = __gnix_rndzv_iov_req_post;
+
+	/* Try posting the txds */
+	ret = __gnix_rndzv_iov_req_post(req);
```

Comment:
You would need to clean up all the TXDs allocated here if you failed to post the requests.  I don't think we'd expect them to fail in healthy network conditions, though